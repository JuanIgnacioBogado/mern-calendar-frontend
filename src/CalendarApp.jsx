import { BrowserRouter } from 'react-router-dom';

import { AppRouter } from './router';

export const CalendarApp = () => {
  return (
    <BrowserRouter>
      <AppRouter />
    </BrowserRouter>
  );
};
