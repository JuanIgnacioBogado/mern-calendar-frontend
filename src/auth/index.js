export * from './pages/LoginPage';

export * from './components/Login';
export * from './components/Register';
