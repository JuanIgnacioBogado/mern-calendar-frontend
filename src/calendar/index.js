export * from './components/CalendarEvent';
export * from './components/CalendarModal';
export * from './components/FabAddNew';
export * from './components/Navbar';

export * from './pages/CalendarPage';
