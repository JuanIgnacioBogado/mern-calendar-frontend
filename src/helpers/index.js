export * from './calendarLocalizer';
export * from './getMessages';
