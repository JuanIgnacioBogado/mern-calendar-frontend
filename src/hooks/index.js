export * from './useAuthStore';
export * from './useCalendarStore';
export * from './useDispatchActions';
export * from './useForm';
export * from './useSelectorStore';
