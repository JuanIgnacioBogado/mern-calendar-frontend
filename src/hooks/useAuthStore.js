import { useMemo } from 'react';

import { calendarApi } from '../api';

import { useSelectorStore } from './';

export const useAuthStore = () => {
  const { auth, onChecking, onLogin, onLogout, clearErrorMessage, clearCalendar } = useSelectorStore();

  const showErrorMessage = (data = 'Credenciales Incorrectas') => {
    if (data?.errors) {
      onLogout(
        Object.values(data?.errors)
          .map(({ msg }) => msg)
          .join('<br><br>')
      );
    } else {
      onLogout(data?.msg || data);
    }
    setTimeout(() => clearErrorMessage(), 10);
  };

  // eslint-disable-next-line no-unused-vars
  const setUserAndToken = ({ token, ok, ...user }) => {
    localStorage.setItem('token', token);
    localStorage.setItem('token-init-date', new Date().getTime());
    onLogin(user);
  };

  const startLogin = async formState => {
    onChecking();
    try {
      const { data } = await calendarApi.post('/auth', formState);

      setUserAndToken(data);
    } catch ({ response: { data } }) {
      console.log('data', data?.msg || data?.errors);
      showErrorMessage();
    }
  };

  const startRegister = async formState => {
    onChecking();
    try {
      const { data } = await calendarApi.post('/auth/new', formState);

      setUserAndToken(data);
    } catch ({ response: { data } }) {
      showErrorMessage(data);
    }
  };

  const checkAuthToken = async () => {
    if (!localStorage.getItem('token')) return startLogout();

    onChecking();
    try {
      const { data } = await calendarApi.get('/auth/renew');

      setUserAndToken(data);
    } catch ({ response: { data } }) {
      showErrorMessage(data);
      startLogout();
    }
  };

  const startLogout = () => {
    localStorage.removeItem('token');
    localStorage.removeItem('token-init-date');
    onLogout();
    clearCalendar();
  };

  const isAuthenticated = useMemo(() => auth.status === 'authenticated', [auth.status]);
  const isChecking = useMemo(() => auth.status === 'checking', [auth.status]);

  return {
    ...auth,
    checkAuthToken,
    isAuthenticated,
    isChecking,
    showErrorMessage,
    startLogin,
    startRegister,
    startLogout
  };
};
