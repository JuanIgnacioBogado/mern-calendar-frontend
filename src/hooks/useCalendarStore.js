import { calendarApi } from '../api';
import { useSelectorStore } from './';

import parseISO from 'date-fns/parseISO';

const parseEvent = ({ start, end, ...event }) => ({ start: parseISO(start), end: parseISO(end), ...event });
const parseUser = (_id, name) => ({ user: { _id, name } });

export const useCalendarStore = () => {
  const {
    auth: {
      user: { uid, name }
    },
    addNewEvent,
    calendar,
    deleteEvent,
    onOpenDateModal,
    onCloseDateModal,
    setActiveEvent,
    setEvents,
    setIsLoading,
    ui,
    updateEvent
  } = useSelectorStore();

  const startGetEvents = async () => {
    setIsLoading(true);
    try {
      const { events } = (await calendarApi.get('/events')).data;
      setEvents(events.map(parseEvent));
    } catch (error) {
      console.log('error', error);
    } finally {
      setIsLoading(false);
    }
  };

  const startNewEvent = async newEvent => {
    setIsLoading(true);
    try {
      const { event } = (await calendarApi.post('/events', newEvent)).data;
      addNewEvent({
        ...parseEvent(event),
        ...parseUser(uid, name)
      });
    } catch (error) {
      console.log('error', error);
    } finally {
      setIsLoading(false);
    }
  };

  const startUpdateEvent = async ({ id, ...updatedEvent }) => {
    setIsLoading(true);
    try {
      const { event } = (await calendarApi.put(`/events/${id}`, updatedEvent)).data;
      updateEvent({
        ...parseEvent(event),
        ...parseUser(uid, name)
      });
    } catch (error) {
      console.log('error', error);
    } finally {
      setIsLoading(false);
    }
  };

  const startDeleteEvent = async id => {
    setIsLoading(true);
    try {
      await calendarApi.delete(`/events/${id}`);
      deleteEvent(id);
    } catch (error) {
      console.log('error', error);
    } finally {
      setIsLoading(false);
    }
  };

  return {
    ...calendar,
    ...ui,
    onOpenDateModal,
    onCloseDateModal,
    setActiveEvent,
    startDeleteEvent,
    startGetEvents,
    startNewEvent,
    startUpdateEvent
  };
};
