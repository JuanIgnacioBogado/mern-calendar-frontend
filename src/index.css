/* Modal */
.ReactModalPortal > div {
  opacity: 0;
}

.ReactModalPortal .ReactModal__Overlay {
  align-items: center;
  display: flex;
  justify-content: center;
  transition: opacity 0.2s ease-in-out;
  z-index: 999;
}

.modal-fondo {
  background-color: rgba(0, 0, 0, 0.3);
  bottom: 0;
  left: 0;
  right: 0;
  top: 0;
  position: fixed;
}

.ReactModalPortal .ReactModal__Overlay--after-open {
  opacity: 1;
}

.ReactModalPortal .ReactModal__Overlay--before-close {
  opacity: 0;
}

.modal {
  background: white;
  border-radius: 10px;
  color: rgb(51, 51, 51);
  display: inline;
  max-height: 620px;
  max-width: 500px;
  outline: none;
  padding: 10px;
  /* height: auto */
}

/* Fabs */
.fab {
  border-radius: 100%;
  bottom: 25px;
  font-size: 25px;
  padding: 15px;
  position: fixed;
  right: 25px;
}
