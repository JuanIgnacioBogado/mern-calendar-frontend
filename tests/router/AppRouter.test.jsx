import { render, screen } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';

import { AppRouter } from '../../src/router';
import { useAuthStore } from '../../src/hooks/useAuthStore';

jest.mock('../../src/hooks/useAuthStore');
jest.mock('../../src/calendar/pages/CalendarPage', () => ({
  CalendarPage: () => <h1>Calendar Page</h1>
}));

describe('AppRouter', () => {
  const mockCheckAuthToken = jest.fn();

  beforeEach(() => jest.clearAllMocks());

  test('should to show loading screen and call checkAuthToken', () => {
    useAuthStore.mockReturnValue({
      isChecking: true,
      checkAuthToken: mockCheckAuthToken
    });

    render(<AppRouter />);

    expect(screen.getByText(/loading/i)).toBeTruthy();
    expect(mockCheckAuthToken).toBeCalled();
  });

  test('should to show login screen if are not authenticated', () => {
    useAuthStore.mockReturnValue({
      isChecking: false,
      isAuthenticated: false,
      checkAuthToken: mockCheckAuthToken
    });

    const { container } = render(
      <MemoryRouter initialEntries={['/']}>
        <AppRouter />
      </MemoryRouter>
    );

    expect(container).toMatchSnapshot();
    expect(screen.getByText(/ingreso/i)).toBeTruthy();
    expect(screen.getByText(/registro/i)).toBeTruthy();
    expect(mockCheckAuthToken).toBeCalled();
  });

  test('should to show calendar screen if are authenticated', () => {
    useAuthStore.mockReturnValue({
      isChecking: false,
      isAuthenticated: true,
      checkAuthToken: mockCheckAuthToken
    });

    render(
      <MemoryRouter initialEntries={['/auth/login']}>
        <AppRouter />
      </MemoryRouter>
    );

    expect(mockCheckAuthToken).toBeCalled();
    expect(screen.getByText(/calendar page/i)).toBeTruthy();
  });
});
