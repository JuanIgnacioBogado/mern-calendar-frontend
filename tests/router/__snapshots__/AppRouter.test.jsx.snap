// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppRouter should to show login screen if are not authenticated 1`] = `
<div>
  <div
    class="container login-container animate__animated animate__fadeIn animate__faster"
  >
    <div
      class="row"
    >
      <div
        class="col-md-6 login-form-1"
      >
        <h3>
          Ingreso
        </h3>
        <form
          class="d-flex flex-column gap-3"
        >
          <input
            class="form-control"
            name="email"
            placeholder="Correo"
            type="email"
            value=""
          />
          <input
            autocomplete="off"
            class="form-control"
            name="password"
            placeholder="Contraseña"
            type="password"
            value=""
          />
          <input
            class="btnSubmit"
            type="submit"
            value="Login"
          />
        </form>
      </div>
      <div
        class="col-md-6 login-form-2"
      >
        <h3>
          Registro
        </h3>
        <form
          class="d-flex flex-column gap-3"
        >
          <input
            class="form-control"
            name="name"
            placeholder="Nombre"
            type="text"
            value=""
          />
          <input
            class="form-control"
            name="email"
            placeholder="Correo"
            type="email"
            value=""
          />
          <input
            autocomplete="off"
            class="form-control"
            name="password"
            placeholder="Contraseña"
            type="password"
            value=""
          />
          <input
            autocomplete="off"
            class="form-control"
            name="password2"
            placeholder="Repita la contraseña"
            type="password"
            value=""
          />
          <input
            class="btnSubmit"
            type="submit"
            value="Crear cuenta"
          />
        </form>
      </div>
    </div>
  </div>
</div>
`;
